File: temps/flutter/lib/common/whileAsync.dart

part of 'FA.dart';

class FakeWhile {
  var mark = 'RUNNING';
  FakeWhile();
  exec(condition, callback) async {
    if (condition()) {
      if (mark == 'RETURN') {
        return 'RETURN';
      }
      if (mark == 'CONTINUE') {
        return exec(condition, callback);
      }
      if (mark == 'BREAK') {
        return 'BREAK';
      }
      await callback(command);

      return await exec(condition, callback);
    } else {
      return 'END';
    }
  }
  command(type) {
    mark = type;
  }
}

File: temps/Web-Vue3/src/common/whileAsync.ts
class FakeWhile {
  mark = 'RUNNING'
  constructor() {
  }
  async exec(condition, callback) {
    if (condition()) {
      if (this.mark == 'RETURN') {
        return 'RETURN'
      }
      if (this.mark == 'CONTINUE') {
        return this.exec(condition, callback)
      }
      if (this.mark == 'BREAK') {
        return 'BREAK'
      }
      await callback(this.command.bind(this))
      return await this.exec(condition, callback)
    } else {
      return 'END'
    }
  }
  command(type: string) {
    this.mark = type
  }
}

export default async function whileAsync(condition, callback) {
  return new FakeWhile().exec(condition, callback)
}
